File: web/src/core/boardData.ts
import { GridPos, PieceType, gridPos } from "./types";

export class BoardData {
  readonly width: number;
  readonly height: number;
  readonly grid: PieceType[][];

  constructor(width: number, height: number) {
    if (width <= 0 || height <= 0) {
      throw new Error("Board dimensions must be positive.");
    }
    this.width = width;
    this.height = height;
    this.grid = Array.from({ length: width }, () =>
      Array.from({ length: height }, () => PieceType.None),
    );
  }

  get(pos: GridPos): PieceType {
    return this.grid[pos.x][pos.y];
  }

  set(pos: GridPos, type: PieceType): void {
    this.grid[pos.x][pos.y] = type;
  }

  isValid(pos: GridPos): boolean {
    return pos.x >= 0 && pos.x < this.width && pos.y >= 0 && pos.y < this.height;
  }

  clone(): BoardData {
    const copy = new BoardData(this.width, this.height);
    for (let x = 0; x < this.width; x++) {
      for (let y = 0; y < this.height; y++) {
        copy.grid[x][y] = this.grid[x][y];
      }
    }
    return copy;
  }
}

export function forEachCell(
  width: number,
  height: number,
  fn: (pos: GridPos) => void,
): void {
  for (let y = 0; y < height; y++) {
    for (let x = 0; x < width; x++) {
      fn(gridPos(x, y));
    }
  }
}

File: Assets/Scripts/Data/BoardData.cs
using System;

namespace Match3.Data
{
    /// <summary>
    /// 棋盘底层数据：二维 PieceType 网格。
    /// </summary>
    public class BoardData
    {
        public int Width { get; }
        public int Height { get; }
        public PieceType[,] Grid { get; }

        public BoardData(int width, int height)
        {
            if (width <= 0 || height <= 0)
                throw new ArgumentOutOfRangeException(nameof(width), "Board dimensions must be positive.");

            Width = width;
            Height = height;
            Grid = new PieceType[width, height];
        }

        public PieceType Get(int x, int y) => Grid[x, y];

        public PieceType Get(GridPos pos) => Get(pos.X, pos.Y);

        public void Set(int x, int y, PieceType type) => Grid[x, y] = type;

        public void Set(GridPos pos, PieceType type) => Set(pos.X, pos.Y, type);

        public bool IsValid(int x, int y) =>
            x >= 0 && x < Width && y >= 0 && y < Height;

        public bool IsValid(GridPos pos) => IsValid(pos.X, pos.Y);

        public BoardData Clone()
        {
            var copy = new BoardData(Width, Height);
            Array.Copy(Grid, copy.Grid, Grid.Length);
            return copy;
        }
    }
}
